tsdb/engine/tsm1/compact.go

```diff
@@ -0,0 +1,298 @@
+package tsm1
+
+// Compactions are the process of creating read-optimized TSM files.
+// The files are created by converting write-optimized WAL entries
+// to read-optimized TSM format.  They can also be created from existing
+// TSM files when there are tombstone records that neeed to be removed, points
+// that were overwritten by later writes and need to updated, or multiple
+// smaller TSM files need to be merged to reduce file counts and improve
+// compression ratios.
+//
+// The the compaction process is stream-oriented using multiple readers and
+// iterators.  The resulting stream is written sorted and chunk to allow for
+// one-pass writing of a new TSM file.
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"sort"
+)
+
+var errMaxFileExceeded = fmt.Errorf("max file exceeded")
+
+// Compactor merges multiple WAL segments and TSM files into one or more
+// new TSM files.
+type Compactor struct {
+	Dir         string
+	MaxFileSize int
+	currentID   int
+
+	merge *MergeIterator
+}
+
+func (c *Compactor) Compact(walSegments []string) ([]string, error) {
+	var walReaders []*WALSegmentReader
+
+	// For each segment, create a reader to iterate over each WAL entry
+	for _, path := range walSegments {
+		f, err := os.Open(path)
+		if err != nil {
+			return nil, err
+		}
+		r := NewWALSegmentReader(f)
+		defer r.Close()
+
+		walReaders = append(walReaders, r)
+	}
+
+	// WALKeyIterator allows all the segments to be ordered by key and
+	// sorted values during compaction.
+	walKeyIterator, err := NewWALKeyIterator(walReaders...)
+	if err != nil {
+		return nil, err
+	}
+
+	// Merge iterator combines the WAL and TSM iterators (note: TSM iteration is
+	// not in place yet).  It will also chunk the values into 1000 element blocks.
+	c.merge = NewMergeIterator(walKeyIterator, 1000)
+	defer c.merge.Close()
+
+	// These are the new TSM files written
+	var files []string
+
+	for {
+		// TODO: this needs to be intialized based on the existing files on disk
+		c.currentID++
+
+		// New TSM files are written to a temp file and renamed when fully completed.
+		fileName := filepath.Join(c.Dir, fmt.Sprintf("%07d.%s.tmp", c.currentID, Format))
+
+		// Write as much as possible to this file
+		err := c.write(fileName)
+
+		// We've hit the max file limit and there is more to write.  Create a new file
+		// and continue.
+		if err == errMaxFileExceeded {
+			files = append(files, fileName)
+			continue
+		}
+
+		// We hit an error but didn't finish the compaction.  Remove the temp file and abort.
+		if err != nil {
+			os.RemoveAll(fileName)
```

Comment:
Well you're ignoring the error. If that is the final plan you could call
Remove() and ignore that error too. :-)

On Tuesday, November 24, 2015, Jason Wilder <notifications@github.com>
wrote:

> In tsdb/engine/tsm1/compact.go
> <https://github.com/influxdb/influxdb/pull/4900#discussion_r45828733>:
>
> > +		// New TSM files are written to a temp file and renamed when fully completed.
> > +		fileName := filepath.Join(c.Dir, fmt.Sprintf("%07d.%s.tmp", c.currentID, Format))
> > +
> > +		// Write as much as possible to this file
> > +		err := c.write(fileName)
> > +
> > +		// We've hit the max file limit and there is more to write.  Create a new file
> > +		// and continue.
> > +		if err == errMaxFileExceeded {
> > +			files = append(files, fileName)
> > +			continue
> > +		}
> > +
> > +		// We hit an error but didn't finish the compaction.  Remove the temp file and abort.
> > +		if err != nil {
> > +			os.RemoveAll(fileName)
>
> It won't return an error if the file happens to be gone (somehow).
> os.Remove returns a *pathError in that case. We don't care about that
> error because we just need the file to be removed.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/influxdb/influxdb/pull/4900/files#r45828733>.
>